### PR TITLE
Geocat-viz package structure (restructure branch)

### PR DESCRIPTION
**Pull request by anissa111**
This PR changes the structure of geocat-viz to resemble other geocat packages such as geocat-comp and geocat-f2py. 

Summary of changes:
- implements pre-commit hooks
- creates a place for future documentation on RTD
- restructures the repo for the purpose of adding plotting functionality that exceeds convenience/utility functions
- adds a conda environment specific to -viz development
- updates to README and INSTALLATION files to reflect new functionality and structure

Notes:
- this PR needs to be merged before further updates can be made to Taylor diagrams and the improvements made to NCL_Plot and Contour can be added
- The package build issues have been resolved, seemingly by a force restart of my machine and me trying to reproduce the error in front of somebody else
- An additional PR should be made for getting something up for documentation, instead of essentially a blank RTD page
- I'm sure I've missed something that needs to be updated
- This PR was absolutely a group effort, thank you @pilotchute and @michaelavs!

### CODE REVIEW

**Review by pilotchute — APPROVED** [review]

**Review by erogluorhan — COMMENTED** [review]
Many files from under "docs/_build/" directory seem to be included in the PR, but I think they are generated files and they should be going into the repository. Is it correct?

**Comment by anissa111** [comment]
> Many files from under "docs/_build/" directory seem to be included in the PR, but I think they are generated files and they should be going into the repository. Is it correct?

Yup, good catch thank you!

**Comment by erogluorhan** [comment]
> Many files from under "docs/_build/" directory seem to be included in the PR, but I think they are generated files and they should be going into the repository. Is it correct?

Typo correction: "... should NOT be going to ..." Sorry

File: docs/index.rst [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+.. .. image:: _static/images/nsf.png
+..    :scale: 100%
+..   :align: right
+
+.. |
+.. |
+
+GeoCAT-viz
+==========
+
```
Comment (erogluorhan) [inline docs/index.rst:15]: Knowing that this docs fodler and index.rst will initiate GeoCAT-viz wevpage but entirely empty, could you at least put a paragraph of the descriptive text about GeoCAT-viz (probably from contributor's guide) here?

File: docs/conda_environment.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+channels:
+  - conda-forge
+  - ncar
+dependencies:
```
Comment (erogluorhan) [inline docs/conda_environment.yml:4]: This file seems to have been copied from geocat-comp and has several unnecessary dependencies for GeoCAT-viz, which should be removed.

File: conda_environment.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+name: geocat-viz
+channels:
+  - conda-forge
+  - ncar
+dependencies:
```
Comment (erogluorhan) [inline conda_environment.yml:5]: In general, such conda_environment files are aimed at "building from source" (for those who won't install via conda, and instead build from source and maybe contribute further with their own development. If we have this file, then:

- Several dependencies listed in it seem like not needed for building/testing purposes (e.g. `metpy`, `sphinx`, etc). 
- Shouldn't we add a "Building from source" part into INSTALLATION.md?

**Review by erogluorhan — CHANGES_REQUESTED** [review]
I have shared some more inline comments as well as:
- Please get rid of test folder and __init__.py in it if it will be empty.

File: readthedocs.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# .readthedocs.yml
```
Comment (erogluorhan) [inline readthedocs.yml:1]: The file name should start with a period I think.

File: meta.yaml [hunk 5]
```diff
@@ -1,26 +1,25 @@
-{% set version = "2020.7.30.1" %}
+{% set version = "2021.07.0" %}
 
 package:
   name: geocat-viz
   version: {{ version }}
 
 source:
-  url: https://github.com/NCAR/geocat-viz/archive/v{{ version }}.tar.gz
+  git_rev: restructure # https://github.com/NCAR/geocat-viz/archive/v{{ version }}.tar.gz
```
Comment (erogluorhan) [inline meta.yaml:8]: I believe we should be using the `main` branch instead of `restructure`.

File: readthedocs.yml (hunk 4 above)
Comment (anissa111) [reply readthedocs.yml]: Yup, thank you!

File: conda_environment.yml (hunk 3 above)
Comment (anissa111) [reply conda_environment.yml:5]: Good point! Added

File: docs/conda_environment.yml (hunk 2 above)
Comment (anissa111) [reply docs/conda_environment.yml:4]: You are absolutely correct on where this file came from, haha. I've modified it now.

File: docs/index.rst (hunk 1 above)
Comment (anissa111) [reply docs/index.rst:15]: That's fair! I stole a paragraph from the README for the moment.

**Comment by anissa111** [comment]
> I have shared some more inline comments as well as:
> 
> * Please get rid of test folder and **init**.py in it if it will be empty.

I can get rid of it for now, but I believe it will be getting content eventually

File: meta.yaml (hunk 5 above)
Comment (anissa111) [reply meta.yaml]: Yup thank you!

File: meta.yaml [hunk 6]
```diff
@@ -1,26 +1,25 @@
-{% set version = "2020.7.30.1" %}
+{% set version = "2021.07.0" %}
 
 package:
   name: geocat-viz
   version: {{ version }}
 
 source:
```
Comment (erogluorhan) [inline meta.yaml:7]: This conda builds might be aimed at either debugging or actual release, so it mwould be great to put descriptive comments for future reference. Something similar to:

```
#git_rev: main   # For building from main branch for debugging purposes
git_tag: v{{ version }}   # For building from a specific tag
git_url: https://github.com/ncar/geocat-viz.git
#url: https://github.com/NCAR/geocat-viz/archive/v{{ version }}.tar.gz   # For building from a specific tag's tarball
```

FYI: Using `git_tag` plus `git_url` above gives us the same way of releasing our actual versions (to be uploaded into conda) as we do in GeoCAT-comp or GeoCAT-f2py, where we first "tag" our `main` branch as a release at any point in the repo, then we generate our conda package from that "tagged" release with this `meta.yaml` file. On the other hand, the developer could anytime use `git_rev` plus `git_url` or only `url` itself for debugging purposes in order to generate a test conda package from any branch or specific tagged release, respectively. We can discuss the details further if you wish.

File: meta.yaml [hunk 7]
```diff
@@ -1,26 +1,25 @@
...
-  url: https://github.com/NCAR/geocat-viz/archive/v{{ version }}.tar.gz
+  git_rev: main # https://github.com/NCAR/geocat-viz/archive/v{{ version }}.tar.gz
+  git_url: https://github.com/ncar/geocat-viz.git
 
 build:
-  noarch: python
```
Comment (erogluorhan) [inline meta.yaml:11]: Is there a specific reason for removing `noarch`?

File: meta.yaml [hunk 8]
```diff
@@ -1,26 +1,25 @@
...
 
 requirements:
   host:
     - python
     - setuptools
     - pip
-    - numpy
   run:
     - python
-    - {{ pin_compatible('numpy') }}
```
Comment (erogluorhan) [inline meta.yaml:23]: I am wondering what this `pin_compatible` was doing before. Any idea?

File: requirements.txt [hunk 9]
```diff
@@ -0,0 +1,4 @@
+matplotlib
```
Comment (erogluorhan) [inline requirements.txt:1]: This `requirements.txt` files is for creating PyPI releases. In order to automate PyPI releases whenever we have a tagged release, you may want to add a pypi.yaml file similar to [GeoCAT-comp's PyPI YAML file](https://github.com/NCAR/geocat-comp/blob/main/.github/workflows/pypi.yaml) unless you have plans of manually releasing PyPI versions.

**Review by erogluorhan — CHANGES_REQUESTED** [review]
It is on a great track! Please address inline comments.

**Comment by erogluorhan** [comment]
> > I have shared some more inline comments as well as:
> > 
> > * Please get rid of test folder and **init**.py in it if it will be empty.
> 
> I can get rid of it for now, but I believe it will be getting content eventually

Oh okay, I thought you ended up deciding not to include any tests. Knowing tests will be created in the future, it is up to you to remove or keep.

File: .gitignore [hunk 10]
```diff
@@ -1,2 +1,7 @@
 .idea/
 .DS_Store
+/docs/_build/
+.ipynb_checkpoints/
+
+# Sphinx documentation
+docs/_build/
```
[the remark this answers is not in the document]
Comment (anissa111) [reply .gitignore:7]: Good catch, thank you!

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,24 +1,48 @@
...
+
+        - Otherwise, you may create and self-assign an issue under
+        [GeoCAT-viz Issues](https://github.com/NCAR/GeoCAT-viz/issues)
         that describes need for the functionality you are planning to contribute.
 
-2. Define a new function and implement it under the `$GEOCAT_VIZ/src/util.py`. 
\ No newline at end of file
+2. Define a new function and implement it under the `$GEOCAT_VIZ/src/geocat/viz/util/util.py`.
+
+# Adding new plotting functionality to the Geocat-viz repo
```
[the remark this answers is not in the document]
Comment (anissa111) [reply CONTRIBUTING.md:26]: Good point, I've altered it to be
```suggestion
# Adding new plotting utility functionality to the Geocat-viz repo
```

File: conda_environment.yml [hunk 12]
```diff
@@ -0,0 +1,18 @@
+name: geocat-viz
```
[the remark this answers is not in the document]
Comment (anissa111) [reply conda_environment.yml:1]: Good point, changed!

File: INSTALLATION.md [hunk 13]
```diff
@@ -21,9 +21,46 @@ compatibility between dependency packages.
...
+
+
+### How to create a Conda environment for building GeoCAT-viz
+
+The GeoCAT-viz source code includes a conda environment definition file,
+`conda_environment.yml` in the root directory that can be used to create a
+development environment containing all of the packages required to build GeoCAT-viz.
+
+    conda env create -f conda_environment.yml
+    conda activate geocat-viz
```
[the remark this answers is not in the document]
Comment (anissa111) [reply INSTALLATION.md:56]: Done!

File: README.md [hunk 14]
```diff
@@ -1,26 +1,21 @@
...
-
-
 # Documentation
 
 [GeoCAT Homepage](https://geocat.ucar.edu/)
 
 [GeoCAT Contributor's Guide](https://geocat.ucar.edu/pages/contributing.html)
 
-[GeoCAT-Examples Gallery](https://geocat-examples.readthedocs.io)
+[GeoCAT-comp documentation on Read the Docs](https://geocat-comp.readthedocs.io)
```
[the remark this answers is not in the document]
Comment (anissa111) [reply README.md:14]: Oh dear, haha, thanks!

File: conda_environment.yml [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  - python=3
+  - geocat-comp
+  - geocat-datafiles
+  - cartopy
+  - make
+  - matplotlib
+  - netcdf4
+  - scikit-learn
+  - sphinx
+  - wrf-python
```
[the remark this answers is not in the document]
Comment (anissa111) [reply conda_environment.yml:15]: Good point! I left those in for testing purposes, but we don't have that yet. I removed them!

File: docs/conda_environment.yml [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+  - python
+  - geocat-datafiles  # for tests
+  - metpy
+  - netcdf4  # for tests
+  - numpy
+  - pytest  # for tests
```
[the remark this answers is not in the document]
Comment (anissa111) [reply docs/conda_environment.yml:10]: Done!

File: meta.yaml (hunk 6 above)
Comment (anissa111) [reply meta.yaml:7]: Oh! I had no idea this was a thing, thanks! I've put your suggestion in.

File: meta.yaml (hunk 7 above)
Comment (anissa111) [reply meta.yaml:11]: Not that I'm aware of. I believe this was a change made by Alea. 

@pilotchute, what does this do?

File: meta.yaml (hunk 8 above)
Comment (anissa111) [reply meta.yaml:23]: Same as above. @pilotchute, any ideas?

File: requirements.txt (hunk 9 above)
Comment (anissa111) [reply requirements.txt:1]: Oh, neat, I've never run the pypi releases. I added a similar file, but I think I need to add PYPI_PASSWORD as a secret, like in comp. How do I do that?

Comment (erogluorhan) [reply requirements.txt:1]: I have created and added the secret "PYPI_GEOCAT_VIZ" to this repo for you. You should just make sure to refer to exact that name within your pypi.yaml file.

FYI: The steps for PyPI configuration is as follows:

1. Create an account on pypi.org
2. Create an API token from https://pypi.org/manage/account/ that will be used to upload the built package. 
3. Add the API token as a secret to your GitHub repo settings. The secret's name must be referred to from with in `pypi.yaml` file in your repo.

File: .github/workflows/pypi.yaml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+          python setup.py sdist bdist_wheel
+          python -m pip wheel . -w dist --no-deps
+      - name: Test the artifacts
+        run: |
+          python -m twine check dist/*
+      - name: Publish package to PyPI
+        uses: pypa/gh-action-pypi-publish@v1.4.2
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI_PASSWORD }}
```
Comment (erogluorhan) [inline .github/workflows/pypi.yaml]: Do not forget to change "PYPI_PASSWORD" to "PYPI_GEOCAT_VIZ"

**Review by erogluorhan — COMMENTED** [review]
It is in almost perfect shape. There is one inline comment and one below:

I gave it a local try, `pip install`ed it into my geocat-examples conda env and then ran `make html` within my GeoCAT-examples; there was no error. However, when I attempt to debug any GeoCAT-examples script from within my PyCharm, my IDE yells at `from . import util` of `src/geocat/viz/__init__.py`.

It may or may not be something we should be concerned about. I will try to dig into it later, or you could do the same.

File: .github/workflows/pypi.yaml (hunk 17 above)
Comment (anissa111) [reply .github/workflows/pypi.yaml]: Ahh yeah, we knew this was going to be a thing. We changed the directory structure, so I think the way we import util functions from viz needs to change. I'm not sure why it's yelling at you from the init in viz, but I think that's what's going on.

The version of geocat-viz is pinned in geocat-examples, though, so this release wouldn't immediately break -examples and we could make the necessary changes before unpinning it.

**Comment by anissa111** [comment]
> It is in almost perfect shape. There is one inline comment and one below:
> 
> I gave it a local try, `pip install`ed it into my geocat-examples conda env and then ran `make html` within my GeoCAT-examples; there was no error. However, when I attempt to debug any GeoCAT-examples script from within my PyCharm, my IDE yells at `from . import util` of `src/geocat/viz/__init__.py`.
> 
> It may or may not be something we should be concerned about. I will try to dig into it later, or you could do the same.

Ahh yeah, we knew this was going to be a thing. We changed the directory structure, so I think the way we import util functions from viz needs to change. I'm not sure why it's yelling at you from the init in viz, but I think that's what's going on.

The version of geocat-viz is pinned in geocat-examples, though, so this release wouldn't immediately break -examples and we could make the necessary changes before unpinning it.

**Comment by erogluorhan** [comment]
> > It is in almost perfect shape. There is one inline comment and one below:
> > I gave it a local try, `pip install`ed it into my geocat-examples conda env and then ran `make html` within my GeoCAT-examples; there was no error. However, when I attempt to debug any GeoCAT-examples script from within my PyCharm, my IDE yells at `from . import util` of `src/geocat/viz/__init__.py`.
> > It may or may not be something we should be concerned about. I will try to dig into it later, or you could do the same.
> 
> Ahh yeah, we knew this was going to be a thing. We changed the directory structure, so I think the way we import util functions from viz needs to change. I'm not sure why it's yelling at you from the init in viz, but I think that's what's going on.
> 
> The version of geocat-viz is pinned in geocat-examples, though, so this release wouldn't immediately break -examples and we could make the necessary changes before unpinning it.

Yes, the version of geocat-viz being pinned in geocat-examples is good. However, I think we should still better solve the namespacing thing in the PR. I am working on it. 

There is an additional confusion/question of mine: Where is `cmaps` being imported now in `geocat-viz`? It was imported in `src/__init__.py` and used in GeoCAT-examples scripts, being imported from `geocat.viz`, but I cannot see it anywhere now. I couldn't. understand how it didn't create a problem when I ran `make html` in GeoCAT-examples though.

**Review by erogluorhan — APPROVED** [review]
Great work!

**Pull request merged**